### DISCUSSION TITLE
How to set GPIO pull-up / pull-down on RPi in REM

### DISCUSSION THREAD

**Discussion by orrious** [tagyoureit/nodejs-poolController#534] [REM]
In REM, when using the RPi GPIOs as inputs, how do you set the pull-up or pull-down resistors build into the RPi?

**Comment by tagyoureit** [comment]
Set the pin direction as input or output and it will set the resistor value.

**Reply by orrious** [reply]
Once you set it as input, you need to be able to select whether you want it to "Float, pull-up, or pull-down"  There is no auto sensing, it will be determined by the end users need.  Currently I need a pull-up and it's not functioning without a physical resistor.

**Reply by rstrouse** [reply]
The REM software interface is cross platform so there is no pull up/down calls in the code you must set them in the config.txt.  There is a BCM utility out there to do it.   I never rely on the internal pull ups/downs and always use external resistors.  Probably because I am reconfiguring pi’s all the time.

**Reply by orrious** [reply]
Looks like you're using `onoff` as the RPi GPIO library in REM.  If you use https://github.com/nebrius/raspi-gpio it would allow you to set pull-up / pull-down.  

I certainly wouldn't want to rely on BCM utility to do it.

**Reply by rstrouse** [reply]
I chose onoff simply because I need this to work on dog and fruit boards.

**Reply by orrious** [reply]
Since we select the board we are using in REM, could the RPi board types use the raspi-gpio lib?

Don't get me wrong, I'm using a physical pull-up for the time being, just curious for my own interest.

**Reply by rstrouse** [reply]
Yeah I dug around in it but the goofiness with the compiled libraries on the hardware side makes for some platform dependencies.  I had to do some gymnastics to get much of this to operate in mock modes.  You do know that there are differences within the RPI boards and OS implementations.   It makes for some really interesting GPIO interaction,

I personally, never rely on the internal pull ups/downs simply because of some experiences in the past.  Forget that a pin had been configured improperly and exported then hook up some button logic and you can say goodbye to the pin.  All that being said if you insist on using them you can set them up in the config.txt file.

https://forums.raspberrypi.com/viewtopic.php?f=117&t=208748
